### PR TITLE
Fixed #22229 -- Added primary key validation to BaseModelFormSet._construct_form().

### PR DESCRIPTION
**Pull request by jdufresne**
<https://code.djangoproject.com/ticket/22229>

### CODE REVIEW

File: tests/model_formsets/tests.py [hunk 1]
```diff
@@ -1655,6 +1655,69 @@ def test_prevent_change_outer_model_and_create_invalid_data(self):
...
+            'form-0-name': 'Charles',
+        }
+        formset = AuthorFormSet(data=data)
+        self.assertFalse(formset.is_valid())
+        self.assertEqual(
+            formset.errors,
+            [{'id': ['Select a valid choice. That choice is not one of the available choices.']}],
+        )
+
+    def test_validation_with_non_existing_id(self):
```
Comment (timgraham) [inline tests/model_formsets/tests.py]: This test is already passing without the changes in this PR so perhaps better to omit or put in a separate commit?
non_existing -> nonexistent

File: tests/model_formsets/tests.py [hunk 2]
```diff
@@ -1655,6 +1655,69 @@ def test_prevent_change_outer_model_and_create_invalid_data(self):
...
 
+    def test_validation_without_id(self):
+        AuthorFormSet = modelformset_factory(Author, fields='__all__')
+        data = {
+            'form-TOTAL_FORMS': '1',
+            'form-INITIAL_FORMS': '1',
+            'form-MAX_NUM_FORMS': '',
+            'form-0-name': 'Charles',
+        }
+        formset = AuthorFormSet(data=data)
```
Comment (timgraham) [inline tests/model_formsets/tests.py]: I think data is usually not passed as a kwarg, at least it's not needed here.

File: django/forms/models.py [hunk 3]
```diff
@@ -578,20 +578,35 @@ def _get_to_python(self, field):
...
-            pk = to_python(pk)
-            kwargs['instance'] = self._existing_object(pk)
-        if i < self.initial_form_count() and 'instance' not in kwargs:
-            kwargs['instance'] = self.get_queryset()[i]
-        if i >= self.initial_form_count() and self.initial_extra:
+        pk_required = False
+        if i < self.initial_form_count():
+            pk_required = True
+            if self.is_bound:
+                pk_key = "%s-%s" % (self.add_prefix(i), self.model._meta.pk.name)
```
Comment (timgraham) [inline django/forms/models.py]: Could switch to single quotes as long as this is being modified.

File: django/forms/models.py [hunk 4]
```diff
@@ -578,20 +578,35 @@ def _get_to_python(self, field):
...
+                try:
+                    pk = self.data[pk_key]
+                except KeyError:
+                    pass
```
Comment (timgraham) [inline django/forms/models.py]: I'd add a comment explaining the circumstances in which this exception happens. (user tampering with data, most likely?)

File: django/forms/models.py [hunk 5]
```diff
@@ -578,20 +578,35 @@ def _get_to_python(self, field):
...
+                else:
+                    pk_field = self.model._meta.pk
```
Comment (timgraham) [inline django/forms/models.py]: I think the `pk_field` variable could be removed here and below.

File: django/forms/models.py [hunk 6]
```diff
@@ -578,20 +578,35 @@ def _get_to_python(self, field):
...
+                    to_python = self._get_to_python(pk_field)
+                    try:
+                        pk = to_python(pk)
+                    except ValidationError:
+                        pass
```
Comment (timgraham) [inline django/forms/models.py]: Add a comment as above?

File: tests/model_formsets/tests.py [hunk 7]
```diff
@@ -1655,6 +1655,69 @@ def test_prevent_change_outer_model_and_create_invalid_data(self):
...
+        self.assertFalse(formset.is_valid())
```
Comment (timgraham) [inline tests/model_formsets/tests.py]: I think the `is_valid()` assertions can be removed. The `formset.errors` assertions will fail if the formset is valid.

File: tests/model_formsets/tests.py (hunk 2 above)
Comment (jdufresne) [inline tests/model_formsets/tests.py]: Passing data as a kwarg is a pattern throughout `tests/model_formsets/tests.py` which is probably where I copied it from. Would you like to see the entire file cleaned up as a separate commit to avoid these types of copy & paste style conflicts?

Grepping, I found the following on master which could also be cleaned up if that is the preferred style:

```
./docs/ref/forms/api.txt:897:        >>> bound_form = ContactForm(data={'subject': 'My Subject'})
./docs/ref/forms/api.txt:1010:        >>> f = ContactForm(data={'message': ''})
./docs/ref/forms/api.txt:1018:        >>> f = ContactForm(data={'message': ''})
./docs/ref/forms/api.txt:1027:        >>> f = ContactForm(data={'message': ''})
./docs/ref/forms/api.txt:1050:        >>> bound_form = ContactForm(data={'subject': 'hi'}, initial=initial)
./docs/releases/1.3.txt:499:    >>> formset = ArticleFormSet(data=None)
./tests/gis_tests/test_geoforms.py:96:        form = PointForm(data={'pt': 'POINT(5 23)'}, initial={'pt': point})
./tests/gis_tests/test_geoforms.py:225:        form = PointForm(data={'p': geom})
./tests/gis_tests/test_geoforms.py:229:        invalid = PointForm(data={'p': 'some invalid geom'})
./tests/gis_tests/test_geoforms.py:234:            self.assertFalse(PointForm(data={'p': invalid.wkt}).is_valid())
./tests/gis_tests/test_geoforms.py:241:        form = PointForm(data={'p': geom})
./tests/gis_tests/test_geoforms.py:247:            self.assertFalse(PointForm(data={'p': invalid.wkt}).is_valid())
./tests/gis_tests/test_geoforms.py:254:        form = LineStringForm(data={'f': geom})
./tests/gis_tests/test_geoforms.py:260:            self.assertFalse(LineStringForm(data={'p': invalid.wkt}).is_valid())
./tests/gis_tests/test_geoforms.py:267:        form = LineStringForm(data={'f': geom})
./tests/gis_tests/test_geoforms.py:273:            self.assertFalse(LineStringForm(data={'p': invalid.wkt}).is_valid())
./tests/gis_tests/test_geoforms.py:280:        form = PolygonForm(data={'p': geom})
./tests/gis_tests/test_geoforms.py:286:            self.assertFalse(PolygonForm(data={'p': invalid.wkt}).is_valid())
./tests/gis_tests/test_geoforms.py:293:        form = PolygonForm(data={'p': geom})
./tests/gis_tests/test_geoforms.py:299:            self.assertFalse(PolygonForm(data={'p': invalid.wkt}).is_valid())
./tests/gis_tests/test_geoforms.py:306:        form = GeometryForm(data={'g': geom})
./tests/gis_tests/test_geoforms.py:312:            self.assertFalse(GeometryForm(data={'g': invalid.wkt}).is_valid())
./tests/gis_tests/test_geoforms.py:326:        form = PointForm(data={'p': geom})
./tests/gis_tests/test_geoforms.py:398:        form = PointForm(data={'p': point})
./tests/gis_tests/test_geoforms.py:407:        form = PointForm(data={'p': point.json})
./tests/forms_tests/tests/tests.py:175:        f = FileForm(data={}, files={'file1': file1}, auto_id=False)
./tests/forms_tests/tests/tests.py:292:        form = ChoiceFieldExclusionForm(data=data, instance=instance)
./tests/forms_tests/tests/test_formsets.py:1041:        formset = FavoriteDrinksFormSet(data=data)
./tests/forms_tests/tests/test_forms.py:2406:        f = FileForm(data={}, files={}, auto_id=False)
./tests/forms_tests/tests/test_forms.py:2414:        f = FileForm(data={}, files={'file1': SimpleUploadedFile('name', b'')}, auto_id=False)
./tests/forms_tests/tests/test_forms.py:2422:        f = FileForm(data={}, files={'file1': 'something that is not a file'}, auto_id=False)
./tests/forms_tests/tests/test_forms.py:2431:        f = FileForm(data={}, files={'file1': SimpleUploadedFile('name', b'some content')}, auto_id=False)
./tests/forms_tests/tests/test_forms.py:2439:        f = FileForm(data={}, files={'file1': file1}, auto_id=False)
./tests/forms_tests/tests/test_forms.py:2843:        form = NameForm(data={'name': ['bad', 'value']})
./tests/forms_tests/tests/test_forms.py:2847:        form = NameForm(data={'name': ['should be overly', 'long for the field names']})
./tests/forms_tests/tests/test_forms.py:2851:        form = NameForm(data={'name': ['fname', 'lname']})
./tests/forms_tests/tests/test_forms.py:2986:        form = JSONForm(data={'json': '{}'})
./tests/foreign_object/test_forms.py:22:        form = FormsTests.ArticleForm(data={'pub_date': str(datetime.date.today())})
./tests/model_forms/tests.py:260:        form = AwardForm(data=data, instance=award)
./tests/model_forms/tests.py:271:        form = CustomWriterForm(data={'name': ''}, instance=obj)
./tests/model_forms/tests.py:773:        form = FieldOverridesByFormMetaForm(data={
./tests/model_forms/tests.py:818:        form = IncompleteCategoryFormWithFields(data={'name': 'some name', 'slug': 'some-slug'})
./tests/model_forms/tests.py:822:        form = IncompleteCategoryFormWithExclude(data={'name': 'some name', 'slug': 'some-slug'})
./tests/model_forms/tests.py:1721:        form = ArticleForm(data={'categories': [category2.pk]}, instance=article)
./tests/model_forms/tests.py:1725:        form = ArticleForm(data={'categories': [category2.pk]}, instance=article)
./tests/model_forms/tests.py:1967:        form = WriterForm(data={'initial-persons': writer_pks})
./tests/model_forms/tests.py:1976:        form = WriterForm(data={})
./tests/model_forms/tests.py:2190:        f = TextFileForm(data={'description': 'Assistance'})
./tests/model_forms/tests.py:2192:        f = TextFileForm(data={'description': 'Assistance'}, files={})
./tests/model_forms/tests.py:2247:        f = TextFileForm(data={'description': 'Assistance'})
./tests/model_forms/tests.py:2285:        form = CFFForm(data={'f': None})
./tests/model_forms/tests.py:2307:        form = PhotoForm(data=data, files=files)
./tests/model_forms/tests.py:2369:        f = ImageFileForm(data={'description': 'Look, it changed'}, instance=instance)
./tests/model_forms/tests.py:2418:        f = OptionalImageFileForm(data={'description': 'Test'})
./tests/model_forms/tests.py:2479:        f = NoExtensionImageFileForm(data={'description': 'Edited image'}, instance=instance)
./tests/model_forms/tests.py:2618:        form = InventoryForm(data=data, instance=core)
./tests/model_formsets/tests.py:180:        formset = AuthorFormSet(data=data, queryset=qs)
./tests/model_formsets/tests.py:232:        formset = AuthorFormSet(data=data, queryset=qs)
./tests/model_formsets/tests.py:302:        formset = AuthorFormSet(data=data, queryset=qs)
./tests/model_formsets/tests.py:328:        formset = AuthorFormSet(data=data, queryset=qs)
./tests/model_formsets/tests.py:363:        formset = AuthorMeetingFormSet(data=data, queryset=AuthorMeeting.objects.all())
./tests/model_formsets/tests.py:476:        formset = PoetFormSet(data=data, queryset=qs)
./tests/model_formsets/tests.py:890:        formset = PoemFormSet(data=data, instance=poet)
./tests/model_formsets/tests.py:1017:        formset = PoemFormSet(data=data, instance=poet)
./tests/model_formsets/tests.py:1563:        formset = FormSet(data=data, instance=author)
./tests/model_formsets/tests.py:1651:        formset = AuthorFormSet(data=data, queryset=Author.objects.filter(id__in=(author.id,)))
./tests/flatpages_tests/test_forms.py:32:        self.assertTrue(FlatpageForm(data=dict(url='/new_flatpage/', **self.form_data)).is_valid())
./tests/flatpages_tests/test_forms.py:33:        self.assertTrue(FlatpageForm(data=dict(url='/some.special~chars/', **self.form_data)).is_valid())
./tests/flatpages_tests/test_forms.py:34:        self.assertTrue(FlatpageForm(data=dict(url='/some.very_special~chars-here/', **self.form_data)).is_valid())
./tests/flatpages_tests/test_forms.py:36:        self.assertFalse(FlatpageForm(data=dict(url='/a space/', **self.form_data)).is_valid())
./tests/flatpages_tests/test_forms.py:37:        self.assertFalse(FlatpageForm(data=dict(url='/a % char/', **self.form_data)).is_valid())
./tests/flatpages_tests/test_forms.py:38:        self.assertFalse(FlatpageForm(data=dict(url='/a ! char/', **self.form_data)).is_valid())
./tests/flatpages_tests/test_forms.py:39:        self.assertFalse(FlatpageForm(data=dict(url='/a & char/', **self.form_data)).is_valid())
./tests/flatpages_tests/test_forms.py:40:        self.assertFalse(FlatpageForm(data=dict(url='/a ? char/', **self.form_data)).is_valid())
./tests/flatpages_tests/test_forms.py:43:        form = FlatpageForm(data=dict(url='no_leading_slash/', **self.form_data))
./tests/flatpages_tests/test_forms.py:50:        form = FlatpageForm(data=dict(url='/no_trailing_slash', **self.form_data))
./tests/flatpages_tests/test_forms.py:57:        form = FlatpageForm(data=dict(url='/no_trailing_slash', **self.form_data))
./tests/flatpages_tests/test_forms.py:64:        FlatpageForm(data=data).save()
./tests/flatpages_tests/test_forms.py:66:        f = FlatpageForm(data=data)
./tests/flatpages_tests/test_forms.py:86:        f = FlatpageForm(data=data, instance=existing)
./tests/flatpages_tests/test_forms.py:98:        f = FlatpageForm(data=data)
./tests/auth_tests/test_forms.py:640:        form = UserChangeForm(data={}, instance=user)
```

**Comment by jdufresne** [comment]
Thanks for the review. I've incorporated all feedback.

I moved the passing test to an earlier commit, as I found this a useful edge case to test when reorganizing the the code.

Thoughts on the comment wording is welcome.

File: tests/model_formsets/tests.py (hunk 2 above)
Comment (timgraham) [inline tests/model_formsets/tests.py]: Doesn't matter much, I guess.
